src/Compilers/CSharp/Test/Emit/CodeGen/UnsafeTests.cs

```diff
@@ -9539,6 +9539,92 @@ .maxstack  2
 ");
         }
 
+        [Fact, WorkItem(49639, "https://github.com/dotnet/roslyn/issues/49639")]
+        public void CompareToNullWithNestedUnconstrainedTypeParameter()
+        {
+            var verifier = CompileAndVerify(@"
+using System;
+unsafe
+{
+    test<int>(null);
+    S<int> s = default;
+    test<int>(&s);
+
+    static void test<T>(S<T>* s)
+    {
+        Console.WriteLine(s == null);
+        Console.WriteLine(s is null);
+    }
+}
+
+struct S<T> {}
+", options: TestOptions.UnsafeReleaseExe, expectedOutput: @"
+True
+True
+False
+False");
+
+            verifier.VerifyIL("<Program>$.<<Main>$>g__test|0_0<T>(S<T>*)", @"
+{
+  // Code size       21 (0x15)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldc.i4.0
+  IL_0002:  conv.u
+  IL_0003:  ceq
+  IL_0005:  call       ""void System.Console.WriteLine(bool)""
+  IL_000a:  ldarg.0
+  IL_000b:  ldc.i4.0
+  IL_000c:  conv.u
+  IL_000d:  ceq
+  IL_000f:  call       ""void System.Console.WriteLine(bool)""
+  IL_0014:  ret
+}
+");
+        }
+
+        [Fact, WorkItem(49639, "https://github.com/dotnet/roslyn/issues/49639")]
+        public void CompareToNullWithPointerToUnmanagedTypeParameter()
+        {
+            var verifier = CompileAndVerify(@"
+using System;
+unsafe
+{
+    test<int>(null);
+    int i = 0;
+    test<int>(&i);
+
+    static void test<T>(T* t) where T : unmanaged
+    {
+        Console.WriteLine(t == null);
+        Console.WriteLine(t is null);
+    }
+}
+", options: TestOptions.UnsafeReleaseExe, expectedOutput: @"
+True
+True
+False
+False");
+
+            verifier.VerifyIL("<Program>$.<<Main>$>g__test|0_0<T>(T*)", @"
+{
+  // Code size       21 (0x15)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldc.i4.0
+  IL_0002:  conv.u
+  IL_0003:  ceq
+  IL_0005:  call       ""void System.Console.WriteLine(bool)""
+  IL_000a:  ldarg.0
+  IL_000b:  ldc.i4.0
+  IL_000c:  conv.u
+  IL_000d:  ceq
+  IL_000f:  call       ""void System.Console.WriteLine(bool)""
+  IL_0014:  ret
+}
+");
+        }
```

Comment:
Can we add a test where we pattern match against a field which is a pointer and verify that this errors if it's done outside of `unsafe` code? The use of fields as pointers and it's interactions with the `unsafe` context has been a trouble area in the past.